velox/functions/lib/SubscriptUtil.h

```diff
@@ -96,6 +97,42 @@ class LookupTable : public LookupTableBase {
   std::unique_ptr<outer_map_t> map_;
 };
 
+// A flat vector of map keys, an index into that vector and an index into
+// the original map keys vector that may have encodings.
+struct MapKey {
```

Comment:
We can add key type/kind as template parameter, then for primitive types we don't need to call the slow virtual version of the comparisons, only need to specialize it for complex types to forward to virtual comparisons.  Then we can merge `primitiveKeyLookupTable_` and `complexKeyHashMap_`.

Reply:
Good call, but I don't think I want to templatize MapKey, since if I used it for a primitive type, we'd have to make an expensive virtual call every time we access an element in baseVector, rather than storing it in memory directly.

But I can change the parameter of LookupTable to be on the NativeType directly rather than the TypeKind, and then use MapKey as the key if it's void and otherwise use MapKey.  Then I can combine primitiveKeyLookupTable_ and complexKeyHashMap_ and still use the primitive key type when possible.  The only additional cost would be having an outer hash map with a single key for complex types, but that should be cheap as there will only every be a single entry.